src/rustzcash.rs

```diff
@@ -191,6 +191,44 @@ pub extern "system" fn librustzcash_merkle_hash(
     write_le(tmp, &mut result[..]);
 }
 
+
+/// Compute Sapling note commitment.
+#[no_mangle]
+pub extern "system" fn librustzcash_sapling_compute_cm(
+    diversifier: *const [c_uchar; 11],
+    pk_d: *const [c_uchar; 32],
+    value: uint64_t,
+    r: *const [c_uchar; 32],
+    result: *mut [c_uchar; 32],
+) -> bool {
+
+    let diversifier = sapling_crypto::primitives::Diversifier(unsafe { *diversifier });
+    let g_d = match diversifier.g_d::<Bls12>(&JUBJUB) {
+        Some(g_d) => g_d,
+        None => return false
+    };
+
+    let pk_d = match edwards::Point::<Bls12, Unknown>::read(&(unsafe { &*pk_d })[..], &JUBJUB) {
+        Ok(p) => p,
+        Err(_) => return false
+    };
+
+    let pk_d = match pk_d.as_prime_order(&JUBJUB) {
+        Some(pk_d) => pk_d,
+        None => return false
+    };
+
+    let r = <Bls12 as JubjubEngine>::Fs::to_uniform(unsafe { &(&*r)[..] });
```

Comment:
This won't work, because `to_uniform` requires the input to be 64 bytes (so that we get approximate uniformity taking the value mod the field size). `r` needs to already be a valid random Jubjub point, which means we need another API to return such values (via `Fs::rand(&mut rng)`).

Reply:
Let's work backwards.  We want a valid rust note so that we can call `note.cm()`.

```
    let note = sapling_crypto::primitives::Note {
        value, g_d, pk_d, r
    };
```
What is r?   It's the commitment randomness and I've seen it defined as this: (what type is this in rust?)
```
        let commitment_randomness: fs::Fs = rng.gen();
```
So from C++ what type should the commitment randomness be?  Currently in C++ 'r' (short for 'rcm') is a 'uint256' which matches the spec.

If 'to_uniform' only takes 64 bytes, does that mean in C++ 'r' should be a 64 byte buffer?  Yet the spec says rcm must be 256 bits.

@daira @str4d @ebfull

Reply:
The spec (section 4.6.2) says that rcm is chosen uniformly on NoteCommit<sup>Sapling</sup>.Trapdoor := F<sub>r<sub>J</sub></sub> (``Fs`` in the code).

Reply:
@bitcartel the problem here is the API, not the types. The `to_uniform` API is documented as taking 64 bytes of Random Oracle output. In sapling-crypto, this is always the output of BLAKE2b (see `hash_to_scalar` in `src/util.rs`), but @ebfull and I decided to split the two components apart so as to not constrain the generic `ToUniform` trait to always require BLAKE2b.

> What is r? It's the commitment randomness and I've seen it defined as this: (what type is this in rust?)
>
>         let commitment_randomness: fs::Fs = rng.gen();

`let commitment_randomness: fs::Fs` specifies that the type is `Fs`. We can set it to the output of `rng.gen()` because Rust can figure out that the random number generator can be used to create a uniformly-random Fs element (via `Fs::rand(&mut rng)`).

> So from C++ what type should the commitment randomness be? Currently in C++ 'r' (short for 'rcm') is a 'uint256' which matches the spec.

In C++, `r` needs to be the serialization of `commitment_randomness`; in this function we need to simply read it into an Fs element (via the `read_fs` functions introduced in #9). This means we need another API to generate `commitment_randomness` and return it in serialized form (via the `write_fs` function introduced in #9).

Reply:
Oh, `to_uniform` is also used by #9, taking the output of `PRFexpand` (which also uses BLAKE2b internally).

Reply:
Note that sometimes we're given the opening of the commitment (e.g. when we're processing a decrypted note plaintext), rather than needing to create a new random one. This could technically work with an API that took a 64-byte ``r`` (because ``to_uniform`` is a no-op if the input is already in range), but that would be confusing and would need to be thoroughly documented. A separate API to return a uniform-random ``Fs`` element seems preferable.